### ISSUE TITLE
Chamfer mesh modifier doesn't work

### ISSUE THREAD

**Issue by Xerios** [mapbox/mapbox-unity-sdk#878]
As discussed during Unite 2018 conference, this part has never worked for me on any city ( mapbox's data tiles ) or custom data tileset ( imported through mapbox studio ).

This bug was reproduced on the demo PCs and on my laptop.

Sorry for this lazy description but at the current state, this is the best description I can give.
I'll provide more info if necessary, although I've already discussed about this with the original developer of this feature at the demo booth, and was told to create create this ticket as a reminder.

**Comment by jordy-isaac** [comment]
@Xerios Can you please include some screenshots and steps to reproduce it when you're able to do so?

**Comment by Xerios** [comment]
This feature has never worked for me or anyone I know, so reproducing it is as simple as these following  steps:

**Step 1:** Open up any of the existing samples ( preferably with building meshes ),
 _or alternatively create a new abstract map and get to a point where you can vizualize buildings in any way._

**Step 2:** Add mesh modifier named "ChamferHeightModifier",
 _or alternatively, create your own "Mapbox/Modifier/Smooth Height for Buildings Modifier" with whatever settings you want._

**Step 3:** Enjoy the stream of errors: 
```
ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at <e1a80661d61443feb3dbdaac88eeb776>:0)
System.ThrowHelper.ThrowArgumentOutOfRangeException () (at <e1a80661d61443feb3dbdaac88eeb776>:0)
System.Collections.Generic.List`1[T].get_Item (System.Int32 index) (at <e1a80661d61443feb3dbdaac88eeb776>:0)
Mapbox.Unity.MeshGeneration.Modifiers.ChamferHeightModifier.Sides (Mapbox.Unity.MeshGeneration.Data.VectorFeatureUnity feature, Mapbox.Unity.MeshGeneration.Data.MeshData meshData, System.Single hf, System.Int32 originalVertexCount) (at Assets/AssetPackages/Mapbox/Unity/MeshGeneration/Modifiers/MeshModifiers/ChamferHeightModifier.cs:71)
Mapbox.Unity.MeshGeneration.Modifiers.ChamferHeightModifier.Run (Mapbox.Unity.MeshGeneration.Data.VectorFeatureUnity feature, Mapbox.Unity.MeshGeneration.Data.MeshData md, Mapbox.Unity.MeshGeneration.Data.UnityTile tile) (at Assets/AssetPackages/Mapbox/Unity/MeshGeneration/Modifiers/MeshModifiers/ChamferHeightModifier.cs:57)
Mapbox.Unity.MeshGeneration.Modifiers.MergedModifierStack.Execute (Mapbox.Unity.MeshGeneration.Data.UnityTile tile, Mapbox.Unity.MeshGeneration.Data.VectorFeatureUnity feature, Mapbox.Unity.MeshGeneration.Data.MeshData meshData, UnityEngine.GameObject parent, System.String type) (at Assets/AssetPackages/Mapbox/Unity/MeshGeneration/Modifiers/MergedModifierStack.cs:138)
Mapbox.Unity.MeshGeneration.Interfaces.VectorLayerVisualizer.Build (Mapbox.Unity.MeshGeneration.Data.VectorFeatureUnity feature, Mapbox.Unity.MeshGeneration.Data.UnityTile tile, UnityEngine.GameObject parent) (at Assets/AssetPackages/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs:336)
Mapbox.Unity.MeshGeneration.Interfaces.VectorLayerVisualizer+<ProcessLayer>d__20.MoveNext () (at Assets/AssetPackages/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs:265)
Mapbox.Unity.Utilities.Runnable+Routine.MoveNext () (at Assets/AssetPackages/Mapbox/Unity/Utilities/Runnable.cs:130)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at C:/buildslave/unity/build/Runtime/Export/Coroutines.cs:17)
```

**Comment by brnkhy** [comment]
Hey @Xerios !
I'll look into this as soon as possible, thanks for the ticket! I'll ping you again when I have a fix!

**Comment by jordy-isaac** [comment]
@Xerios Thanks for posting this and my apologies for the hassle. It helps to have this documented in a public ticket if the original convo happened offline at the conf. The context might help others users and devs. I appreciate it :)

**Comment by brnkhy** [comment]
@Xerios I just realized we actually pushed a new version of this (with small changes and probably the fix you needed) 1-2 weeks ago. I just tried it and worked for me;
![aspowerbar_2018-06-24_01-24-38](https://user-images.githubusercontent.com/1465390/41814143-d28d2300-774d-11e8-954f-e6634e724938.png)

latest on dev branch is here;
https://github.com/mapbox/mapbox-unity-sdk/blob/develop/sdkproject/Assets/Mapbox/Unity/MeshGeneration/Modifiers/MeshModifiers/ChamferHeightModifier.cs

can you try it out and let me know if it doesn't work agian please?

**Comment by Xerios** [comment]
Did a quick test using your settings with my maps, and it looks like it works!
![image](https://user-images.githubusercontent.com/181384/41818472-1975cd78-77b0-11e8-85b3-a50a8203029d.png)

Although, it's a bit unfortunate that it doesn't work when going through the simple Polygon Extrusion, because now I have to modify `ChamferHeightModifier` to make it use a custom height property. 
I hope those two workflows will be unified at some point in the future.

Thanks!

**Comment by brnkhy** [comment]
@Xerios to explain how it works in the background; all options (other than custom) under extrusion type is just a predefined set of modifiers. i.e. "polygon" is just `polygon mesh mod + uv mesh mod + textured height mod` etc etc.
So when you select polygon and then add a chamfer height mod, it applies it after the textured mesh mod . maybe we can make it to override in the future but I'm not sure if that would be intuitive.

As another note, as you can see in your shot, you'll have those tile edge issues using chamfer modifier. Using 3d buildings tileset normally fixes that problem but there is a small issue with that, custom inspector is lacking an important checkbox for that at the moment. 
I believe @atripathi-mb fixed that and it's on dev branch as well. Not in the released package though so you'll need to pull that fix.

**Comment by brnkhy** [comment]
@Xerios I'll close this ticket for now as it looks like it's working but just hard to use (the setup etc). Feel free to open it up again if you ahve any follow up questions though.

**Comment by KSemenenko** [comment]
Hi @Xerios @brnkhy  👋
I also want to get this functionality now, but it doesn't work the same way for me as it does for you.
Maybe there is a chance that you still have a test project?
Or maybe you have some settings other than those on the screenshot?

**Comment by Xerios** [comment]
@KSemenenko I've posted this nearly 5 years ago. Unfortunately since this was part of a project from my previous job, I did not have the rights to retain the code after I left.

I advise you to create a **new issue** and explain in detail to the core developers what you mean by "doesn't work the same way for me as it does for you", perhaps with a screenshot example of what you want to achieve. I wish you best of luck!
